Show transactions as not fully confirmed during background validation

Sjors opened the pull request:
Fixes #28598: currently if assumeutxo were to be used, transactions might get shown as confirmed even while the entire chainstate is unconfirmed.

This also adds an `assumed` value to `listtransactions` etc, shown only for confirmed transactions when a snapshot is loaded.

The UI experience for assume utxo still needs work. This PR makes the wallet UI safer by not rendering transactions as fully confirmed until the background validation is done.

Instead it shows the same icon as when a transaction has 1 confirmation. This is quite conservative. The cost of tricking someone into accepting a fake coin, by giving them a invalid snapshot`*` and then mining a block on top of it, is higher than a regular doublespend by mining.

![pending](https://github.com/bitcoin/bitcoin/assets/10217/a2a95190-a1a7-402e-822b-b0829d5bdf45)

`* =` in the current implementation this involves giving someone a malicious binary or getting them to recompile

Caveat: because transaction state is updated after `blockConnected`, if you restart the node during background sync, then the transaction will show as fully confirmed (`"assumed": false`) briefly - until the next block is connected.

Sjors committed:
wallet: track background validation height

Sjors committed:
wallet: add IsTxAssumed() to WalletTxStatus

Determine if a given transaction belongs to a block that is assumed to be valid pending background validation.

Add this information to WalletTxStatus.

Sjors commented:
There's a glitch in this implementation though, probably not a big deal in practice. Because it relies on `blockConnected` to update the icon, if you restart the node during background sync, then the transaction will show as fully confirmed briefly - until the next block is connected. Similarly, once background sync finishes, it will show as partially confirmed until a new block comes in.

Sjors commented:
> If someone is running a malicious binary or malicious code

They'd just add code to sweep the wallet.

There's also the (remote) possibility of Bitcoin Core shipping an otherwise valid binary with just a bad hash (accidentally or otherwise).

I think it's reasonable to have _some_ UI differentiator between before and after the background check is finished. There's a difference between your node having validated every block and other people having vouched for the blocks. That could also be a warning message (which would be followed by a crash if background sync completes and finds a different UTXO set).

Sjors commented on src/wallet/wallet.cpp:
```diff
@@ -1468,12 +1468,30 @@ void CWallet::transactionRemovedFromMempool(const CTransactionRef& tx, MemPoolRe
     assert(block.data);
     LOCK(cs_wallet);
 
+    switch (role) {
```
be2be21d1740966325213258151d7f3477b19e1e: if #28608 lands first, this would check `role.validated` and `role.most_work` instead (cc @ryanofsky)

pablomartin4btc commented on src/wallet/wallet.cpp:
```diff
@@ -1468,12 +1468,30 @@ void CWallet::transactionRemovedFromMempool(const CTransactionRef& tx, MemPoolRe
+            break;
+    } // no default case, so the compiler can warn about missing cases
+
+    if (role == ChainstateRole::BACKGROUND) {
+        m_background_validation_height = block.height;
+        return;
+    } else if (role == ChainstateRole::NORMAL) {
+        m_background_validation_height = -1;
+    } else {}
```
Conditions were covered by the `switch` above.
```suggestion
```

Sjors replied:
Dropped.

pablomartin4btc reviewed (commented):
Concept ACK

It makes sense to me to add more information regarding `assumeutxo` when it's enabled or in used, I'm wondering if, apart from the GUI we should also specify it in the rpc command `listtransactions` output.

pablomartin4btc commented on src/wallet/wallet.cpp:
```diff
@@ -3376,7 +3376,6 @@ bool CWallet::IsTxAssumed(const CWalletTx& wtx) const EXCLUSIVE_LOCKS_REQUIRED(c
     return false;
 }
 
-
```
On the`gui:` comit fff5cb42dfe91ce1ba1a1aa2269253f262c6e2b4, was this unintentional?

Sjors replied:
Dropped

pablomartin4btc reviewed (commented):
Small suggestion, maybe the `if`/ `else` style format change on `src/qt/transactionrecord.cpp` in the `gui:` commit fff5cb42dfe91ce1ba1a1aa2269253f262c6e2b4 should be done on separated or previous commit, so it'd be easier to review the change in the logic (setting the new `AssumedConfirmed` status).

Sjors commented:
>> Until you receive coins, there's no risk anyway.
> Can you explain this?

The only thing a GUI wallet user can do is generate an address. This is safe whether or not IBD and background sync are complete, and regardless of if the snapshot hash was valid. Once they receive coins on that address it's useful to know if these coins are in their mempool, confirmed in a block but pending background sync, or confirmed in a fully validated chain.

> attach a wallet over P2P to a node that has no wallet compiled in

There's no p2p message to communicate background validation state.

A wallet that's connected over IPC (see #19461) should figure out it's own handling of this. The way process seperation is designed at the moment, I believe you can only connect the GUI to a remote Node + Wallet combo, so this PR would work. I.e. you can connect a GUI + wallet to a remote node, that would require rethinking the interface.

I also don't think it's a good idea to make the GUI do more independent reasoning, e.g. taking confirmation height from the wallet interface, the background sync state from the node interface and then determining what to display. 

> Are you aware of the RPC call to get the state, which can be accessed via the UI.

I assume we'll add a GUI menu option to load the snapshot soon enough. At that point relying on users calling an RPC method and interpreting the result, doesn't seem like a good idea. It's also safer to not rely on the user proactively checking if the background sync is done. Rather with this PR, they will see that the transaction is not fully confirmed, and if they're curious they can look at that RPC method.

> putting it in the wallet doesn't seem right.

I think we need to access background sync information in two places:
1. The GUI sync progress indicator(s) 
2. The wallet

(1) is independent of the wallet and not implemented in this PR. I don't think it would be sufficient. We should not expect users to automatically understand that a background sync in progress has implications for a confirmed transaction near the tip.

Sjors commented:
> also specify it in the rpc command `listtransactions` output

I'll look into that.

Sjors commented:
I would expect that a wallet that connects to us over (untrusted) p2p will get the exact same info regardless of background validation progress.

Sjors commented on src/wallet/rpc/transactions.cpp:
```diff
@@ -22,16 +22,23 @@ static void WalletTxToJSON(const CWallet& wallet, const CWalletTx& wtx, UniValue
     } else {
         entry.pushKV("trusted", CachedTxIsTrusted(wallet, wtx));
+        // This can happen when loading a wallet from another node,
+        // while still in (background) IBD:
```
82f19c7e97281bbd66a25e53f30fe2f80a4f9d11: I found this while loading an existing wallet into a fresh (signet) node, while it was loading doing the background sync. That's not a very common scenario, so I haven't bothered looking into deeper causes...

Sjors commented on test/functional/feature_assumeutxo.py:
```diff
@@ -12,6 +12,9 @@
 - TODO: test submitting a transaction and verifying it appears in mempool
+- TODO: the "assumed" field for a confirmed wallet transaction should be
+        present during background sync, start false and become true as
+        sync progresses
```
82f19c7e97281bbd66a25e53f30fe2f80a4f9d11: such a test seems hard to write without a (background sync aware) `stopatheight` RPC. So far I just tested manually that the value is correct: absent before loading a snapshot, `true` while background sync is below it, `false` once background sync passed the confirmation height.

Sjors commented:
There's no way for a node to tell its peers that the mempool transactions its relaying rely on an assumed state. The same goes for block relay. Are you suggesting we should change the p2p protocol to add this information?

Sjors commented on test/functional/feature_assumeutxo.py:
```diff
@@ -12,6 +12,9 @@
 - TODO: test submitting a transaction and verifying it appears in mempool
+- TODO: the "assumed" field for a confirmed wallet transaction should be
+        present during background sync, start false and become true as
+        sync progresses
```
Interesting that you're getting that error when loading from a backup, but not when loading a regular wallet.

Sjors replied:
> I think wallet tests should be separate from consensus tests.

That makes sense.

Sjors replied:
Added a wallet test file here: 59796c3926cb146229a2b3e942a5f6994f27d72a

It made me realise we should probably disallow loading a wallet with a birth date below the assume valid block. Though for manual testing it's handy.

Sjors replied:
I'm still trying to wrap my head around the logic in `CWallet::AttachChain` which is what triggers the error when loading a backup, but somehow doesn't trigger it when loading an existing wallet.

Sjors replied:
Oh I see, the test creates the backup _before_ the snapshot height.

Sjors replied:
As for the behavior that this PR introduces, ideally there would a `stopatheight` RPC, but the test could also - as it does now - use `-stopatheight` and restart the node. Since it won't automatically connect to the other test nodes, we can inspect the result of `getwallettransactions` and see if `assumed` is set correctly.

I'll work on that and make it a separate commit, so it's a bit easier to extract the test scaffold if this PR doesn't make it.

Sjors commented:
Rebased to see if the asserts added in #28546 catch anything here. (They don't in the test suite and manual testing).

pablomartin4btc reviewed (commented):
Since you moved the `wallet_assumeutexo.py` out into #28838 perhaps you can remove commit a68f1596b358cfed0b80334ef7429d1c5475aa4c from this PR now.

My previous comments regarding some parts of the code were not addressed nor answered if you could please take a look when's possible.

Sjors commented:
@pablomartin4btc I'll wait for that PR to get merged and then I'll go through review comments again.

fjahr commented:
> > `* =` in the current implementation this involves giving someone a malicious binary or getting them to recompile
> 
> Not sure if this pull request is needed then.
> 
> If someone is running a malicious binary or malicious code, the malicious actor will have this code removed from the binary or code.

I have been thinking about this concept for a while now and I tend to agree that this is not necessary for similar reasons as mentioned above but I am also not against this being added. However, I don't consider this a blocker for assumeutxo mainnet params. So ~0.

Another question for me is: Why don't we have something similar when `assumevalid` is used? (We don't, to my knowledge)

Sjors commented:
Rebased, but not much else changed given the lack of concept-enthusiasm.

Test commit is dropped since #28838 landed.

> I don't consider this a blocker for assumeutxo mainnet params

Agreed. If anywhere, it's a blocker for having _automatic_ and/or very easy UI based snapshot loading.

> Another question for me is: Why don't we have something similar when `assumevalid` is used? 

That state is permanent, since we don't have a background check for `assumevalid`.

> Small suggestion, maybe the `if`/ `else` style format change on `src/qt/transactionrecord.cpp` in the `gui:` ... should be done on separated or previous commit

Going to leave that for now, though I agree the change if a bit noisy.

luke-jr commented on src/qt/transactionrecord.cpp:
```diff
@@ -175,42 +175,31 @@ void TransactionRecord::updateStatus(const interfaces::WalletTxStatus& wtx, cons
     if (type == TransactionRecord::Generated) {
-        if (wtx.blocks_to_maturity > 0)
-        {
+        if (wtx.blocks_to_maturity > 0) {
```
These lines shouldn't be reformatted for no reason

Sjors replied:
I'm changing this function, which seems like the right time for it. Per developer guidelines:

> When writing patches, favor the new style over attempting to mimic the surrounding style

luke-jr commented on src/qt/transactiontablemodel.cpp:
```diff
@@ -320,6 +320,9 @@ QString TransactionTableModel::formatTxStatus(const TransactionRecord *wtx) cons
         status = tr("Abandoned");
         break;
+    case TransactionStatus::AssumedConfirmed:
+        status = tr("%1 confirmations, pending verification of historical blocks").arg(wtx->status.depth);
```
It's not really the historical blocks being verified as much as it is the snapshot...

Maybe just

```suggestion
        status = tr("%1 confirmations, pending verification").arg(wtx->status.depth);
```

luke-jr commented on src/qt/transactiontablemodel.cpp:
```diff
@@ -465,6 +468,8 @@ QVariant TransactionTableModel::txStatusDecoration(const TransactionRecord *wtx)
         return QIcon(":/icons/transaction_0");
```
These are weaker too. Normally, there needs to be a valid transaction to even show up - but in this case, it would be pretty cheap to fake it (no need to mine a block, just get them to use the wrong UTXO set).

Sjors replied:
Do you mean a fake UTXO set which contains a transaction that pays to the user wallet? Presumably that wouldn't show as `Unconfirmed`, but in this PR it would show as `AssumedConfirmed`.

Are you suggesting we should hide it entirely? Or some other state?

luke-jr replied:
For example, an attacker might give a lot of people a UTXO set with one additional entry, controlled by his key. Then, he can at any point later send that UTXO to anyone who used his modified UTXO set.

Sjors replied:
@luke-jr that doesn't work, because we'll refuse to load it. The attacker would have to give them a malicious Bitcoin Core binary, but such a binary can just sweep the wallet.

GBKS commented:
> This PR makes the wallet UI **safer** by not rendering transactions as fully confirmed until the background validation is done.

Could you please clarify what you mean by **safer**? Are there specific scenarios you are thinking of?

From a user perspective, the big benefit of assumeUTXO is that they can use their wallet more quickly. If we then show transactions but communicate to users that they should not really trust what is shown, it does not feel like the wallet is actually ready, which defeats the purpose of the feature. But I might be misunderstanding something, so I'd appreciate the clarification. Thanks in advance.

Sjors commented:
> it does not feel like the wallet is actually ready, which defeats the purpose of the feature

It _isn't_ ready, but it's usable. Before assumeutxo, when a user creates a new wallet and receives coins while the blockchain is still downloading, they would not see those coins. Now they do see the coins, and they can spend them.

But until the background validation is complete, there is a different trust assumption. It's safer than an unconfirmed transaction, because we can check the proof-of-work. It's about the same as an SPV wallet, which only checks the headers and a merkle inclusion proof.

Only if you fully trust that the development process prevented a malicious / buggy snapshot from being included, can you treat it as fully equivalent to the scenario where you completed background validation. But in that case, you might as well not run the background validation at all - yet we do. So I think we should make these transactions (somehow) look a little less final.

GBKS commented:
> So I think we should make these transactions (somehow) look a little less final.

Thanks for the response. My thinking is really around using the "1 confirmation" icon. A transaction having 1 confirmation does not communicate very closely that the transaction has 7, but that those 7 have not been fully verified. Another treatment could be to show the full verification icon, but grey it out. 

We'll also have to resolve this for the [Bitcoin Core App](https://bitcoincore.app/milestones/1-5-activity/).

Sjors commented:
Rebased.

> Another treatment could be to show the full verification icon, but grey it out.

That seems reasonable, but I'm not sure how to implement it. It would need to be a flag that's applied to both the `TransactionStatus::Confirming` and `TransactionStatus::Confirmed` states. Patch is welcome, but otherwise it could be improved in a followup.

> There are reasons for the background download.

I agree there are additional benefits to doing the background sync, beyond verification.

Although this:

> populate the header tree structure with metadata, such as the transaction count of a block, or the transaction count of the chain

... could be committed to separately - it's not consensus critical and overkill to have to download an entire block for.

Downloading historical historical blocks is primarily useful to validate them, secondarily to serve them other nodes (for the same reason not everyone should run a pruned node).

> may encourage to increase the resource usage

In this case I assume you're referring to future protocol changes that would encourage that? Because within the current protocol use (some) people already don't feel much restraint unfortunately. I suspect _time_ is the worst enemy, with linear growth over centuries being the biggest barrier to background sync. Unless Moore's Law doesn't stop and continues to offset this linear growth with exponential performance increase in bandwidth, CPU and I/O (Utreexo only fixes the I/O bottleneck).

GBKS commented:
Not related to the review of this PR, but I thought I'd follow up on my previous comment (let me know if you don't want me to do this). For the Bitcoin Core App, I am proposing to have show the verification status in the text field that we also use for the date and other statuses in transactions. Users can always click a transaction to see the date if they really need to. This keeps the list simple to parse, which is especially important on mobile. The general verification status is indicated in the mini block clock in the top-right (which also has a helpful tooltip), and the larger block clock when you navigate to that screen. More [here](https://bitcoincore.app/snapshot/), and all feedback is appreciated.

![image](https://github.com/bitcoin/bitcoin/assets/695901/df60040f-3e37-4ad4-b164-20de29c018a8)

GBKS commented:
Was just thinking that an alternative approach could be to leave the transactions list as is, but show a message in the send screen that explains the current status and trust assumptions. This is much more explicit than slightly tweaked icons. Just wanted to add it as an idea for an alternative approach, feel free to ignore for the purpose of this PR.

Sjors committed:
gui: add assumed confirmed state

Create a separate status for transactions that are confirmed in
a block that is assumed valid pending background validation.

Use the same icon as for transactions with a single confirmation.

Sjors commented:
Rebased and added release note.

@maflcko once the UI supports loading a snapshot, this PR should be in the same release (or earlier). But there's no PR implementing that yet, so no specific release to target.

I don't think there's any outstanding feedback (minor tweaks in iconography and wording can wait for a followup). Some people find it useful, others don't.

fanquake commented:
Not sure what to do here. Seems unlikely this is going to go into `29.x`, and as far as I'm aware, nobody is working on: ["once the UI supports loading a snapshot"](https://github.com/bitcoin/bitcoin/pull/28616#issuecomment-2385586088) in https://github.com/bitcoin-core/gui. Reading the thread, it doesn't really seem like there is buy in for this change from other contributors or reviewers.

Sjors commented:
> nobody is working on

There is ongoing work in the GUI-QML repo to have snapshot import https://github.com/bitcoin-core/gui-qml/pull/424

But that project is going to take a while, so I'll just make this up for grabs for now.
